gems/smithy-client/spec/smithy-client/param_validator_spec.rb

```diff
@@ -0,0 +1,317 @@
+        it 'accepts TrueClass and FalseClass' do
+          validate(boolean: true)
+          validate(boolean: false)
+          validate({ boolean: 'true' },
+                   'expected params[:boolean] to be true or false, got class String instead.')
+        end
+      end
+
+      describe 'bytes' do
+        it 'accepts a byte' do
+          validate(byte: 123)
+          validate({ byte: '123' },
+                   'expected params[:byte] to be an Integer, got class String instead.')
+        end
+      end
+
+      describe 'documents' do
+        it 'accepts numeric objects' do
+          validate(document: 123)
+          validate(document: 3.14159)
+        end
+
+        it 'accepts string objects' do
+          validate(document: 'test string')
+        end
+
+        it 'accepts booleans' do
+          validate(document: true)
+          validate(document: false)
+        end
+
+        it 'accepts nil' do
+          validate(document: nil)
+        end
+
+        it 'rejects Objects' do
+          expect = 'expected params[:document] to be one of Hash, Array, Numeric, String, TrueClass, ' \
+                   'FalseClass, NilClass, got class Object instead'
+          validate({ document: Object.new }, [expect])
+        end
+
+        describe 'Hash' do
+          it 'accepts a flat Hash' do
+            validate(document: { a: 1, b: 2 })
+          end
+
+          it 'accepts a nested Hash' do
+            validate(document: { a: 1, b: { c: 'nested' } })
+          end
+
+          it 'recursively validates and rejects a nested arbitrary Object' do
+            expect = 'expected params[:document][b][c] to be one of Hash, Array, Numeric, String, ' \
+                     'TrueClass, FalseClass, NilClass, got class Object instead'
+            validate({ document: { a: 1, b: { c: Object.new } } }, [expect])
+          end
+        end
+
+        describe 'Array' do
+          it 'accepts an array' do
+            validate(document: [1, 2, 3])
+          end
+
+          it 'rejects an array with arbitrary Objects' do
+            expect = 'expected params[:document] to be one of Hash, Array, Numeric, String, TrueClass, ' \
+                     'FalseClass, NilClass, got class Object instead'
+            validate({ document: [1, 2, Object.new] }, [expect])
+          end
+        end
+      end
+
+      describe 'doubles' do
+        it 'accepts a double' do
+          validate(double: 123.0)
+          validate({ double: 123 },
+                   'expected params[:double] to be a Float, got class Integer instead.')
+        end
+      end
+
+      describe 'enums' do
+        it 'accepts a string' do
+          validate(enum: 'abc')
+          validate({ enum: 123 },
+                   'expected params[:enum] to be a String, got class Integer instead.')
+        end
+      end
+
+      describe 'floats' do
+        it 'accepts a float' do
+          validate(float: 123.0)
+          validate({ float: 123 },
```

Comment:
This validation feels weird to me - it seems like we should accept an integer here?

Reply:
In v3 we do both param validation and param conversion. Param validation should ensure strictness-ish and param conversion can convert integers given to float shapes as floats. This makes error messaging cleaner maybe. Unless you have a different idea.